src/librustc_typeck/collect.rs

```diff
@@ -48,9 +48,48 @@ use rustc::hir::GenericParamKind;
 use rustc::hir::{self, CodegenFnAttrFlags, CodegenFnAttrs, Unsafety};
 
 use std::iter;
+use std::ops::Range;
 
 struct OnlySelfBounds(bool);
 
+pub fn lint_duplicate_auto_traits<'a, 'gcx, 'tcx>(
+    tcx: TyCtxt<'a, 'gcx, 'tcx>,
+    node_id: ast::NodeId,
+    auto_traits: impl IntoIterator<Item = (Option<Ty<'tcx>>, DefId, Span)>,
```

Comment:
If you make the binder changes below, this would be `Option<Binder<Ty<'tcx>>>`, but I think the code would otherwise work precisely the same.